src/provider/soft.rs

```diff
@@ -0,0 +1,912 @@
+            }
+            RS256Key::Tpm { .. } => Err(TpmError::IncorrectKeyType),
+        }
+    }
+
+    fn rs256_unseal_data(
+        &mut self,
+        key: &RS256Key,
+        sealed_data: &SealedData,
+    ) -> Result<Zeroizing<Vec<u8>>, TpmError> {
+        match (key, sealed_data) {
+            (
+                RS256Key::SoftAes256GcmV2 {
+                    key: _,
+                    content_encryption_key,
+                },
+                SealedData::SoftV1 { data, tag, iv },
+            ) => {
+                unwrap_aes256gcm_nonce16!(content_encryption_key, data, tag, iv)
+            }
+            (
+                RS256Key::SoftAes256GcmV2 {
+                    key: _,
+                    content_encryption_key,
+                },
+                SealedData::SoftAes256GcmV2 { data, tag, nonce },
+            ) => {
+                unwrap_aes256gcm!(content_encryption_key, data, tag, nonce)
+            }
+            (RS256Key::Tpm { .. }, _) | (_, SealedData::TpmAes256GcmV2 { .. }) => {
+                Err(TpmError::IncorrectKeyType)
+            }
+        }
+    }
+}
+
+impl TpmMsExtensions for SoftTpm {
+    fn rs256_oaep_dec_sha1(
+        &mut self,
+        rs256_key: &RS256Key,
+        encrypted_data: &[u8],
+    ) -> Result<Zeroizing<Vec<u8>>, TpmError> {
+        match rs256_key {
+            RS256Key::SoftAes256GcmV2 { key, .. } => {
+                let padding = rsa::Oaep::new::<sha1::Sha1>();
+                key.decrypt(padding, encrypted_data)
+                    .map(|data| data.into())
+                    .map_err(|_| TpmError::RsaOaepDecrypt)
+            }
+            RS256Key::Tpm { .. } => Err(TpmError::IncorrectKeyType),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::SoftTpm;
+    use crate::authvalue::AuthValue;
+    use crate::provider::{Tpm, TpmHmacS256, TpmMsExtensions, TpmRS256};
+    use crate::structures::{
+        LoadableHmacS256Key, LoadableRS256Key, LoadableStorageKey, SealedData,
+    };
+    use crypto_glue::traits::Zeroizing;
+
+    #[test]
+    fn soft_tpm_storage() {
+        let soft_tpm = SoftTpm::default();
+
+        crate::tests::test_tpm_storage(soft_tpm);
+    }
+
+    #[test]
+    fn soft_tpm_hmac() {
+        let soft_tpm = SoftTpm::default();
+
+        crate::tests::test_tpm_hmac(soft_tpm);
+    }
+
+    #[test]
+    fn soft_tpm_ecdsa_p256() {
+        let soft_tpm = SoftTpm::default();
+
+        crate::tests::test_tpm_ecdsa_p256(soft_tpm);
+    }
+
+    #[test]
+    fn soft_tpm_rs256() {
+        let soft_tpm = SoftTpm::default();
+
+        crate::tests::test_tpm_rs256(soft_tpm);
+    }
+
+    #[test]
+    fn soft_tpm_msoapxbc() {
+        let soft_tpm = SoftTpm::default();
+
+        crate::tests::test_tpm_msoapxbc(soft_tpm);
+    }
+
+    #[test]
+    fn test_legacy_hmac_load() {
+        let _ = tracing_subscriber::fmt::try_init();
+
+        let auth_value = AuthValue::from([
```

Comment:
Where'd the test values come from? Could be worth documenting just for future folks validating things